ppx/src/reactjs_jsx_ppx.ml

```diff
@@ -1,15 +1,88 @@
 open Ppxlib
-open Ast_helper
-open Asttypes
-open Parsetree
 
-let rec find_opt p = function
-  | [] -> None
-  | x :: l -> if p x then Some x else find_opt p l
+module Builder = struct
+  (* Ast_builder.Default assigns attributes to be the empty.
+     This wrapper re-exports all used fns with attrs arg to override them. *)
+
+  include Ast_builder.Default
+
+  let pexp_apply ~loc ?(attrs = []) e args =
+    let e = Ast_builder.Default.pexp_apply ~loc e args in
+    { e with pexp_attributes = attrs }
+
+  let pexp_ident ~loc ?(attrs = []) e =
+    let e = Ast_builder.Default.pexp_ident ~loc e in
+    { e with pexp_attributes = attrs }
+
+  let value_binding ~loc ~attrs ~pat ~expr =
+    let vb = Ast_builder.Default.value_binding ~loc ~pat ~expr in
+    { vb with pvb_attributes = attrs }
+end
+
+(* [merlinHide] tells merlin to not look at a node, or at any of its
+   descendants. *)
+let merlinHideAttrs =
+  [
+    {
+      attr_name = { txt = "merlin.hide"; loc = Location.none };
+      attr_payload = PStr [];
+      attr_loc = Location.none;
+    };
+  ]
+
+let merlinFocus =
+  {
+    attr_name = { loc = Location.none; txt = "merlin.focus" };
+    attr_payload = PStr [];
+    attr_loc = Location.none;
+  }
 
 let nolabel = Nolabel
 let labelled str = Labelled str
 let optional str = Optional str
+
+module Binding = struct
+  (* Binding is the interface that the ppx uses to interact with the bindings.
+     Here we define the same APIs as the bindings but it generates Parsetree *)
+  module React = struct
+    let null ~loc =
+      Builder.pexp_ident ~loc { loc; txt = Ldot (Lident "React", "null") }
+
+    let array ~loc children =
+      Builder.pexp_apply ~loc
+        (Builder.pexp_ident ~loc
+           { txt = Longident.Ldot (Lident "React", "array"); loc })
+        [ (nolabel, children) ]
+
+    let componentLike ~loc props return =
+      Ptyp_constr
+        ( { loc; txt = Ldot (Lident "React", "componentLike") },
```

Comment:
This and `ReactDOM.domProps` are types, not bindings. But I guess it is ok :)

Reply:
I guess it's fine, right?

Maybe Bindings.Types.* is an overkill?